src/NHibernate/Engine/Query/QueryPlanCache.cs

```diff
@@ -95,26 +95,30 @@ public IQueryExpressionPlan GetHQLQueryPlan(IQueryExpression queryExpression, bo
 			return plan;
 		}
 
-		public FilterQueryPlan GetFilterQueryPlan(string filterString, string collectionRole, bool shallow, IDictionary<string, IFilter> enabledFilters)
+		public IQueryExpressionPlan GetFilterQueryPlan(string filterString, string collectionRole, bool shallow, IDictionary<string, IFilter> enabledFilters)
 		{
-			var key = new FilterQueryPlanKey(filterString, collectionRole, shallow, enabledFilters);
-			var plan = (FilterQueryPlan) planCache[key];
+			return GetFilterQueryPlan(new StringQueryExpression(filterString), collectionRole, shallow, enabledFilters);
+		}
+
+		public IQueryExpressionPlan GetFilterQueryPlan(IQueryExpression queryExpression, string collectionRole, bool shallow, IDictionary<string, IFilter> enabledFilters)
+		{
+			var key = new FilterQueryPlanKey(queryExpression.Key, collectionRole, shallow, enabledFilters);
+			var plan = (IQueryExpressionPlan) planCache[key];
 
 			if (plan == null)
 			{
 				if (log.IsDebugEnabled)
 				{
-					log.Debug("unable to locate collection-filter query plan in cache; generating (" + collectionRole + " : "
-							  + filterString + ")");
+					log.Debug(string.Format("unable to locate collection-filter query plan in cache; generating ({0} : {1})", collectionRole, queryExpression.Key));
```

Comment:
I am used to rarely need the explicit check, because I am mostly using `Format` suffixed methods when having to format a log. Of course sometime we need to format with computed values, or to have a more complex message construction. Then I use the explicit check. I am sensible to this because log messages are almost bloat code in my view, and I rather keep them as compact as possible. (Actually I tend to only log exceptions. I fully agree with [Jeff Atwood blog](https://blog.codinghorror.com/the-problem-with-logging/) on this.)